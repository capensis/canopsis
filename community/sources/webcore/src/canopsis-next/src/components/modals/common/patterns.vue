<template lang="pug">
  v-form(@submit.prevent="submit")
    modal-wrapper(close)
      template(#title)
        span {{ title }}
      template(#text)
        patterns-form(
          v-model="form",
          :name="config.name",
          :entity="config.entity",
          :alarm="config.alarm",
          :event="config.event",
          :total-entity="config.totalEntity"
        )
      template(#actions)
        v-btn(
          :disabled="submitting",
          depressed,
          flat,
          @click="$modals.hide"
        ) {{ $t('common.cancel') }}
        v-btn.primary(
          :disabled="isDisabled",
          :loading="submitting",
          type="submit"
        ) {{ $t('common.submit') }}
</template>

<script>
import { cloneDeep } from 'lodash';

import { MODALS } from '@/constants';

import { modalInnerMixin } from '@/mixins/modal/inner';
import { submittableMixinCreator } from '@/mixins/submittable';
import { confirmableModalMixinCreator } from '@/mixins/confirmable-modal';

import PatternsForm from '@/components/forms/patterns-form.vue';

import ModalWrapper from '../modal-wrapper.vue';

export default {
  name: MODALS.patterns,
  $_veeValidate: {
    validator: 'new',
  },
  components: {
    PatternsForm,
    ModalWrapper,
  },
  mixins: [
    modalInnerMixin,
    submittableMixinCreator(),
    confirmableModalMixinCreator(),
  ],
  data() {
    return {
      form: this.modal.config.patterns ? cloneDeep(this.modal.config.patterns) : {},
    };
  },
  computed: {
    title() {
      return this.config.title || this.$tc('common.pattern', 2);
    },
  },
  methods: {
    async submit() {
      const isFormValid = await this.$validator.validateAll();

      if (isFormValid) {
        if (this.config.action) {
          await this.config.action(this.form);
        }

        this.$modals.hide();
      }
    },
  },
};
</script>
