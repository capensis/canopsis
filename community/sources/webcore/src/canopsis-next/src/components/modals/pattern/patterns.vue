<template lang="pug">
  v-form(@submit.prevent="submit")
    modal-wrapper(close)
      template(slot="title")
        span {{ title }}
      template(slot="text")
        c-patterns-field(
          v-model="form",
          :alarm="config.alarm",
          :entity="config.entity",
          :event="config.event"
        )
      template(slot="actions")
        v-btn(
          :disabled="submitting",
          depressed,
          flat,
          @click="$modals.hide"
        ) {{ $t('common.cancel') }}
        v-btn.primary(
          :disabled="isDisabled",
          :loading="submitting",
          type="submit"
        ) {{ $t('common.submit') }}
</template>

<script>
import { cloneDeep } from 'lodash';

import { MODALS } from '@/constants';

import { createSubmittableMixin } from '@/mixins/submittable';
import { createConfirmableModalMixin } from '@/mixins/confirmable-modal';
import { createValidationErrorsMixin } from '@/mixins/form/validation-errors';

import ModalWrapper from '../modal-wrapper.vue';

export default {
  name: MODALS.patterns,
  $_veeValidate: {
    validator: 'new',
  },
  components: { ModalWrapper },
  mixins: [
    createSubmittableMixin(),
    createValidationErrorsMixin(),
    createConfirmableModalMixin(),
  ],
  data() {
    return {
      form: this.modal.config.patterns ? cloneDeep(this.modal.config.patterns) : {},
    };
  },
  computed: {
    title() {
      return this.config.title || this.$t('modals.patterns.title');
    },
  },
  methods: {
    async submit() {
      const isFormValid = await this.$validator.validateAll();

      if (isFormValid) {
        try {
          if (this.config.action) {
            await this.config.action(this.form);
          }

          this.$modals.hide();
        } catch (err) {
          this.setFormErrors(err);
        }
      }
    },
  },
};
</script>
