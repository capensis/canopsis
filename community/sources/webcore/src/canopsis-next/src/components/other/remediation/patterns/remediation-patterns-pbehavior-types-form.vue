<template lang="pug">
  div
    v-layout(row)
      pbehavior-type-field(
        v-field="form.active_on_pbh",
        :label="$t('remediationPatterns.tabs.pbehaviorTypes.fields.activeOnTypes')",
        :is-item-disabled="isActiveItemDisabled",
        chips,
        multiple
      )
    v-layout(row)
      pbehavior-type-field(
        v-field="form.disabled_on_pbh",
        :label="$t('remediationPatterns.tabs.pbehaviorTypes.fields.disabledOnTypes')",
        :is-item-disabled="isDisabledItemDisabled",
        chips,
        multiple
      )
</template>

<script>
import PbehaviorTypeField from '@/components/other/pbehavior/calendar/partials/pbehavior-type-field.vue';

export default {
  components: { PbehaviorTypeField },
  model: {
    prop: 'form',
    event: 'input',
  },
  props: {
    form: {
      type: Object,
      default: () => ({}),
    },
  },
  methods: {
    isActiveItemDisabled(item) {
      return this.form.disabled_on_pbh.includes(item._id);
    },

    isDisabledItemDisabled(item) {
      return this.form.active_on_pbh.includes(item._id);
    },
  },
};
</script>
