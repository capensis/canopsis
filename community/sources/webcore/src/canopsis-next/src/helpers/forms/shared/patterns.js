/**
 * Default condition for removing patterns
 *
 * @param {Array|undefined} value
 * @returns {boolean}
 */
export function conditionForRemovingPatterns(value) {
  return !value || !value.length;
}

/**
 * Get conditions for removing empty patterns by unsetSeveralFieldsWithConditions
 *
 * @param {Array} keys
 * @param {Function} condition
 * @returns {Object}
 */
export function getConditionsForRemovingEmptyPatterns(
  keys = ['alarm_patterns', 'entity_patterns', 'event_patterns'],
  condition = conditionForRemovingPatterns,
) {
  return keys.reduce((acc, key) => {
    acc[key] = condition;

    return acc;
  }, {});
}
