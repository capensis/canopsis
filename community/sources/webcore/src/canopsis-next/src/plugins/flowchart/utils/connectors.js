import { range } from 'lodash';

import uid from '@/helpers/uid';

/**
 * @typedef {Object} ConnectorPoint
 * @property {number} x
 * @property {number} y
 * @property {string} _id
 */

/**
 * Generate connector
 *
 * @param {number} x
 * @param {number} y
 * @returns {ConnectorPoint}
 */
export const generateConnectorPoint = (x, y) => ({
  x,
  y,
  _id: uid(),
});

/**
 * Generate connector points by rectangle
 *
 * @param {Rect} rect
 * @param {number} count
 * @returns {ConnectorPoint[]}
 */
export const generateConnectorsByRect = (rect, count) => {
  const { width, height, rx, ry, x, y } = rect;
  const sideConnectorsCount = Math.round(count / 4);
  const sideSpacesCount = sideConnectorsCount + 1;
  const startX = x + rx;
  const startY = x + ry;

  const widthConnectorOffset = (width - rx * 2) / (sideSpacesCount);
  const heightConnectorOffset = (height - ry * 2) / (sideSpacesCount);

  const sideConnectors = range(1, sideConnectorsCount + 1);

  const topConnectors = sideConnectors.map(index => generateConnectorPoint(
    startX + widthConnectorOffset * index,
    y,
  ));

  const rightConnectors = sideConnectors.map(index => generateConnectorPoint(
    x + width,
    startY + heightConnectorOffset * index,
  ));

  const bottomConnectors = sideConnectors.map(index => generateConnectorPoint(
    startX + widthConnectorOffset * index,
    y + height,
  ));

  const leftConnectors = sideConnectors.map(index => generateConnectorPoint(
    x,
    startY + heightConnectorOffset * index,
  ));

  const connectors = [
    ...topConnectors,
    ...rightConnectors,
    ...bottomConnectors,
    ...leftConnectors,
  ];

  if (!rx && !ry) {
    /* Angles connectors */
    connectors.push(
      generateConnectorPoint(x, y),
      generateConnectorPoint(x + width, y),
      generateConnectorPoint(x + width, y + height),
      generateConnectorPoint(x, y + height),
    );
  }

  return connectors;
};
