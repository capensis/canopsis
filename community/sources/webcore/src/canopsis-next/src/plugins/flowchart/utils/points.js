/**
 * Calculate points with ghosts points
 *
 * @param {Array} points
 * @returns {Array}
 */
export const getPointsWithGhosts = points => points.reduce((acc, point, index) => {
  const nextIndex = index + 1;
  const nextPoint = points[nextIndex];

  acc.push({ ...point, index });

  if (nextPoint) {
    acc.push({
      x: (point.x + nextPoint.x) / 2,
      y: (point.y + nextPoint.y) / 2,
      ghost: true,
      index: nextIndex,
    });
  }

  return acc;
}, []);
