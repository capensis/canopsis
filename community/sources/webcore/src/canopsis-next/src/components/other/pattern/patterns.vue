<template>
  <patterns-list
    :options.sync="options"
    :patterns="patterns"
    :total-items="patternsMeta.total_count"
    :pending="patternsPending"
    @edit="$listeners.edit"
    @remove="$listeners.remove"
    @remove-selected="$listeners['remove-selected']"
  />
</template>

<script>
import { localQueryMixin } from '@/mixins/query-local/query';
import { entitiesPatternsMixin } from '@/mixins/entities/pattern';

import PatternsList from './patterns-list.vue';

export default {
  components: {
    PatternsList,
  },
  mixins: [
    localQueryMixin,
    entitiesPatternsMixin,
  ],
  mounted() {
    this.fetchList();
  },
  methods: {
    fetchList() {
      return this.fetchPatternsList({ params: this.getQuery() });
    },
  },
};
</script>
