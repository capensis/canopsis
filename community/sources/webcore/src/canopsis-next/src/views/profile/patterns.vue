<template lang="pug">
  div
    c-page-header {{ $t('pattern.patterns') }}
    v-layout(row, wrap)
      v-flex(xs12)
        v-card.ma-2
          v-tabs(
            v-if="hasReadAnyCorporatePatternAccess",
            v-model="activeTab",
            slider-color="primary",
            fixed-tabs
          )
            v-tab(:href="`#${$constants.PATTERN_TABS.patterns}`") {{ $t('pattern.myPatterns') }}
            v-tab-item(:value="$constants.PATTERN_TABS.patterns")
              patterns(
                @edit="showEditPatternModal",
                @remove="showDeletePatternModal",
                @remove-selected="showDeleteSelectedPatternsModal"
              )
            v-tab(:href="`#${$constants.PATTERN_TABS.corporatePatterns}`") {{ $t('pattern.corporatePatterns') }}
            v-tab-item(:value="$constants.PATTERN_TABS.corporatePatterns", lazy)
              corporate-patterns(
                @edit="showEditPatternModal",
                @remove="showDeletePatternModal",
                @remove-selected="showDeleteSelectedPatternsModal"
              )
          patterns(
            v-else,
            @edit="showEditPatternModal",
            @remove="showDeletePatternModal",
            @remove-selected="showDeleteSelectedPatternsModal"
          )

    c-fab-expand-btn(@refresh="refresh", :has-access="hasAccessToCreatePattern")
      v-tooltip(top)
        v-btn(
          slot="activator",
          color="grey darken-1",
          fab,
          dark,
          small,
          @click.stop="showCreatePbehaviorPatternModal"
        )
          v-icon pause
        span {{ createPbehaviorTitle }}
      v-tooltip(top)
        v-btn(
          slot="activator",
          color="blue",
          fab,
          dark,
          small,
          @click.stop="showCreateEntityPatternModal"
        )
          v-icon perm_identity
        span {{ createEntityTitle }}
      v-tooltip(top)
        v-btn(
          slot="activator",
          color="error",
          fab,
          dark,
          small,
          @click.stop="showCreateAlarmPatternModal"
        )
          v-icon notification_important
        span {{ createAlarmTitle }}
</template>

<script>
import { MODALS, PATTERN_TABS, PATTERN_TYPES } from '@/constants';

import { localQueryMixin } from '@/mixins/query-local/query';
import { entitiesPatternsMixin } from '@/mixins/entities/pattern';
import { entitiesCorporatePatternsMixin } from '@/mixins/entities/pattern/corporate';
import {
  permissionsTechnicalProfileCorporatePatternMixin,
} from '@/mixins/permissions/technical/profile/corporate-pattern';

import Patterns from '@/components/other/pattern/patterns.vue';
import CorporatePatterns from '@/components/other/pattern/corporate-patterns.vue';

export default {
  components: {
    Patterns,
    CorporatePatterns,
  },
  mixins: [
    localQueryMixin,
    entitiesPatternsMixin,
    entitiesCorporatePatternsMixin,
    permissionsTechnicalProfileCorporatePatternMixin,
  ],
  data() {
    return {
      activeTab: PATTERN_TABS.patterns,
    };
  },
  computed: {
    isCorporatePatternsTab() {
      return this.activeTab === PATTERN_TABS.corporatePatterns;
    },

    hasAccessToCreatePattern() {
      return !this.isCorporatePatternsTab || this.hasCreateAnyCorporatePatternAccess;
    },

    createAlarmTitle() {
      return this.isCorporatePatternsTab
        ? this.$t('modals.createCorporateAlarmPattern.create.title')
        : this.$t('modals.createAlarmPattern.create.title');
    },

    createEntityTitle() {
      return this.isCorporatePatternsTab
        ? this.$t('modals.createCorporateEntityPattern.create.title')
        : this.$t('modals.createEntityPattern.create.title');
    },

    createPbehaviorTitle() {
      return this.isCorporatePatternsTab
        ? this.$t('modals.createCorporatePbehaviorPattern.create.title')
        : this.$t('modals.createPbehaviorPattern.create.title');
    },
  },
  methods: {
    refresh() {
      if (this.isCorporatePatternsTab) {
        return this.fetchCorporatePatternsListWithPreviousParams();
      }

      return this.fetchPatternsListWithPreviousParams();
    },

    showCreateAlarmPatternModal() {
      this.$modals.show({
        name: MODALS.createPattern,
        config: {
          pattern: { is_corporate: this.isCorporatePatternsTab },
          title: this.createAlarmTitle,
          type: PATTERN_TYPES.alarm,
          action: async (pattern) => {
            await this.createPattern({ data: pattern });

            return this.refresh();
          },
        },
      });
    },

    showCreateEntityPatternModal() {
      this.$modals.show({
        name: MODALS.createPattern,
        config: {
          pattern: { is_corporate: this.isCorporatePatternsTab },
          title: this.createEntityTitle,
          type: PATTERN_TYPES.entity,
          action: async (pattern) => {
            await this.createPattern({ data: pattern });

            return this.refresh();
          },
        },
      });
    },

    showCreatePbehaviorPatternModal() {
      this.$modals.show({
        name: MODALS.createPattern,
        config: {
          pattern: { is_corporate: this.isCorporatePatternsTab },
          title: this.createPbehaviorTitle,
          type: PATTERN_TYPES.pbehavior,
          action: async (pattern) => {
            await this.createPattern({ data: pattern });

            return this.refresh();
          },
        },
      });
    },

    getEditPatternModalTitle(pattern) {
      if (pattern.is_corporate) {
        return {
          [PATTERN_TYPES.alarm]: this.$t('modals.createCorporateAlarmPattern.edit.title'),
          [PATTERN_TYPES.entity]: this.$t('modals.createCorporateEntityPattern.edit.title'),
          [PATTERN_TYPES.pbehavior]: this.$t('modals.createCorporatePbehaviorPattern.edit.title'),
        }[pattern.type];
      }

      return {
        [PATTERN_TYPES.alarm]: this.$t('modals.createAlarmPattern.edit.title'),
        [PATTERN_TYPES.entity]: this.$t('modals.createEntityPattern.edit.title'),
        [PATTERN_TYPES.pbehavior]: this.$t('modals.createPbehaviorPattern.edit.title'),
      }[pattern.type];
    },

    showEditPatternModal(editablePattern) {
      this.$modals.show({
        name: MODALS.createPattern,
        config: {
          pattern: editablePattern,
          type: editablePattern.type,
          title: this.getEditPatternModalTitle(editablePattern.type),
          action: async (pattern) => {
            await this.updatePattern({ id: editablePattern._id, data: pattern });

            return this.refresh();
          },
        },
      });
    },

    showDeletePatternModal(id) {
      this.$modals.show({
        name: MODALS.confirmation,
        config: {
          action: async () => {
            await this.removePattern({ id });

            return this.refresh();
          },
        },
      });
    },

    showDeleteSelectedPatternsModal(selected) {
      this.$modals.show({
        name: MODALS.confirmation,
        config: {
          action: async () => {
            await this.bulkRemovePatterns({
              data: selected.map(({ _id: id }) => ({ _id: id })),
            });

            return this.refresh();
          },
        },
      });
    },
  },
};
</script>
