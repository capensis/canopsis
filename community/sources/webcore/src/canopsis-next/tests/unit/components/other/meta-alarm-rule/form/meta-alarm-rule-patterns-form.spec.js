import { generateEntityPatternsTests } from '@unit/utils/patterns';

import MetaAlarmRulePatternsForm from '@/components/other/meta-alarm-rule/form/meta-alarm-rule-patterns-form.vue';

generateEntityPatternsTests(MetaAlarmRulePatternsForm, 'meta-alarm-rule-patterns-form', { disabled: true });
