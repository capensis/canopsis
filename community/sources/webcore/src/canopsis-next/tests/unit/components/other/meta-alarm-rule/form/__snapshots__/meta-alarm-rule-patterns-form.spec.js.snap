// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`meta-alarm-rule-patterns-form Renders \`meta-alarm-rule-patterns-form\` with custom props 1`] = `
<c-patterns-field-stub
  alarm-attributes="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
  both-counters="both-counters"
  disabled="true"
  entity-attributes="[object Object]"
  value="[object Object]"
  with-alarm="with-alarm"
  with-entity="with-entity"
/>
`;

exports[`meta-alarm-rule-patterns-form Renders \`meta-alarm-rule-patterns-form\` with default props 1`] = `
<c-patterns-field-stub
  alarm-attributes="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
  both-counters="both-counters"
  entity-attributes="[object Object]"
  value="[object Object]"
  with-alarm="with-alarm"
  with-entity="with-entity"
/>
`;
